Add gitlab ci testing tutorial (#1273)
localstack/docs: content/en/tutorials/gitlab_ci_testcontainers/index.md

---

> title: "End-to-End Testing in Gitlab CI with Testcontainers and LocalStack: Understanding Runners and Docker in Docker"

Reviewer: I would drop the last part of the title to keep it simple & crisp. It should be okay to have this info in the description

Author: This is done on purpose, as the main pain points of running on GitLab are the runners and the docker in docker setup. Additionally, all tutorial titles are pretty long, so it's not sticking out from an aesthetic point of view.

linkTitle: "End-to-End Testing in Gitlab CI with Testcontainers and LocalStack: Understanding Runners and Docker in Docker"

> weight: 3

Reviewer: We already have conflicting weights here. Please assign it another weight

description: >
  In this tutorial, we'll walk through the process of setting up end-to-end testing for a backend application using Testcontainers and LocalStack within GitLab. We'll understand the types of GitLab Runners available for CI pipelines and how the concept of Docker-in-Docker plays a crucial role in this environment.
type: tutorials

> teaser: "Learn how you can set up end-to-end testing using Testcontainers and LocalStack in GitLab CI. This tutorial covers setting up the .gitlab-ci.yml file, configuring GitLab runners, and setting up a local Docker runner. With this setup, you can ensure your application is thoroughly tested in a CI/CD environment."

Reviewer: Please indent filenames

Author: indent? you mean use quotation marks?

Reviewer: Sorry, I meant format

- aws-java-sdk
- aws-cli
tags:
- Java
- AWS Java SDK
- GitLab
- GitLab Runners
- Docker-in-Docker
pro: true
leadimage: "ls-gitlab-testcontainers.png"
---

## Introduction: Testcontainers & LocalStack

Testcontainers is an open-source framework that provides lightweight APIs for bootstrapping local development and test dependencies
with real services wrapped in Docker containers.
Running tests with Testcontainers and LocalStack is crucial for AWS-powered applications because it ensures each test runs in a clean,
isolated environment, providing consistency across all development and CI machines. LocalStack avoids AWS costs by emulating
services locally, preventing exceeding AWS free tier limits, and eliminates reliance on potentially unstable external AWS services.
This allows for the simulation of difficult-to-reproduce scenarios, edge cases, and enables testing of the
entire application stack in an integrated manner. Testing with LocalStack and Testcontainers also integrates
seamlessly with CI/CD pipelines like GitLab CI or GitHub Actions, allowing developers to run automated tests without requiring AWS credentials or services.

## Prerequisites

For this tutorial, you will need:

> - [LocalStack Pro](https://localstack.cloud/pricing/) to emulate the AWS services. If you don't have a subscription yet, you can just get a trial license for free.

Reviewer: Please link to the installation doc instead of the pricing page and add `LOCALSTACK_AUTH_TOKEN` as well

- [Docker](https://docker.io/)
- [A GitLab account](https://gitlab.com/)

## Why GitLab?

> GitLab is striving to be a complete tool for DevOps practices, offering not just source code management and continuous integration, but also features for

Reviewer: I think lets be marketing-neutral to third-party services :)

Reviewer: I would keep a very formal introduction: 

> GitLab is a web-based DevOps platform that combines git repository management, continuous integration/continuous deployment (CI/CD) pipelines, and issue tracking. It provides a single application for the entire software development lifecycle.

Reviewer: On a second thought, I would suggest to remove this entire section since in none of the existing tutorials, we had a section like "Why X?". Its on users to find value.

Author: This is a valid paragraph that sets the tone for the entire tutorial. It states facts and explains how GitLab stands out compared to other tools. I'll remove the last sentence so it looks less like we have favorites.

Reviewer: I would still suggest to avoid selling GitLab since its a proprietary service. A user who would like forward to the tutorial will anyway have GitLab set and ready and won't need any push from us to adopt that.

#### Self-hosted runners

Essentially, the architecture does not change, except the runners will be executing the jobs on a local machine. For developing locally, 
this approach is very convenient and there are several benefits:

- **Customization**: you can configure the runners to suit your specific needs and environment. 
- **Performance**: improved performance and faster builds by leveraging your own hardware. 
- **Security**: enhanced control over your data and build environment, reducing exposure to external threats. 
- **Resource Management**: better management and allocation of resources to meet your project's demands.
- **Cost Efficiency**: depending on your alternatives, you can avoid usage fees associated with cloud-hosted runners.

## Application Overview

Our sample backend application stores information about different types of coffee in files, with descriptions stored in an S3 bucket. It utilizes two
Lambda functions to create/update and retrieve these descriptions, all accessible through an API Gateway. While we won't delve
into the details of creating these AWS resources, we'll use AWS CLI to initialize them during container startup using init hooks. You can

> find the whole setup in the [init-resources.sh](https://gitlab.com/tinyg210/coffee-backend-localstack/-/blob/main/src/test/resources/init-resources.sh?ref_type=heads) file.

Reviewer: Please migrate this to https://gitlab.com/localstack.cloud, since maintaining repos on personal accounts/namespaces is not a good precedent

Reviewer: @whummer Can you help @tinyg210 with access to the GitLab repo?

Author: i'll swap links when i get access to it

The following diagram visually explains the simple workflows that we want to check in our automated test in CI, using Testcontainers.
We'll need to make sure that the files are correctly created and named, that the validations and exceptions happen as expected.
<br>
<br>
{{< figure src="coffee-app-diagram.png" width="80%" height="auto">}}

## CI Pipeline Using GitLab Runners

> #### Configuring the test container

Reviewer: H3 tags should follow H2 ones. Not H4 per se.

> To follow along, make changes to the code or run your own pipelines, you can fork the repository at https://gitlab.com/tinyg210/coffee-backend-localstack.

Reviewer 2: ```suggestion
To follow along, make changes to the code or run your own pipelines, you may fork the repository from the [coffee-backend-localstack sample](https://gitlab.com/tinyg210/coffee-backend-localstack).
```

                            MountableFile.forHostPath("target/lambda.jar", 0777),
                            "/etc/localstack/init/ready.d/target/lambda.jar")
                    .withCopyFileToContainer(
                            MountableFile.forClasspathResource("init-resources.sh", 0777),
                            "/etc/localstack/init/ready.d/init-resources.sh")
                    .withEnv("DEBUG", "1")
                    .withNetworkAliases("localstack")
                    .withEnv("LAMBDA_DOCKER_FLAGS", testcontainersLabels())
                    .withEnv("LAMBDA_RUNTIME_ENVIRONMENT_TIMEOUT", "90")
                    .waitingFor(Wait.forLogMessage(".*Finished creating resources.*\\n", 1));
    
    ......
```

Here's a breakdown of what's important:

- The `@Testcontainers` marks the test class to use the Testcontainers library.
- The `@Container` annotation indicates that the field is a Testcontainers managed container.
- The image used for the test LocalStack instance is set to the latest Pro version (at the time of writing).
- In order to use the Pro image, a `LOCALSTACK_AUTH_TOKEN` variable needs to be set and read from the environment.

> - There are two files copied to the container before startup: the jar file for the Lambda functions and the script for building

Reviewer: ```suggestion
- There are two files copied to the container before startup: the JAR file for the Lambda functions and the script for provisioning
```

all the necessary AWS resources. Both files are copied with read/write/execute permissions.

> - `DEBUG=1` enables a more verbose debugging of LocalStack.

Reviewer: ```suggestion
- `DEBUG=1` enables a more verbose logging of LocalStack.
```

- The last line `.waitingFor(Wait.forLogMessage(...))` configures the container to wait until the specified log message appears, exactly once, indicating that resource creation is complete.

{{< alert title="Sidenote" >}}

Ryuk is a component of Testcontainers that helps manage and clean up Docker resources created during testing. Specifically, Ryuk
ensures that any Docker containers, networks, volumes, and other resources are properly removed when they are no longer needed.
This prevents resource leaks and ensures that the testing environment remains clean and consistent between test runs.

When Testcontainers starts, it typically launches a Ryuk container in the background. This container continuously monitors
the Docker resources created by Testcontainers and removes them once the test execution is complete or if they are no longer in use.
{{< /alert >}}

The tests are set up in the `CoffeeAppTests` class, validating the workflows for creating a coffee description files, retrieving them, and exception throwing when needed.
For this tutorial you don't really need to dive into the specifics of the tests, but you're more than welcome to.

> #### Setting up the pipeline configuration

Reviewer: Same goes here and other areas

we need to run Docker commands in our CI/CD jobs, we must configure GitLab Runner to support those commands. 
This method requires `privileged` mode.

Let's use the current running container to do that. Run the following:

```commandline
$ docker exec -it gitlab-runner bin/bash
```

Inside the container, let's run:

{{<command>}}
$ cd etc/gitlab-runner
$ ls -al
<disable-copy>
total 24
drwx------ 3 root root 4096 May 16 19:58 .
drwxr-xr-x 1 root root 4096 May 16 19:56 ..
-rw------- 1 root root   14 May 16 19:50 .runner_system_id
drwx------ 2 root root 4096 May  3 17:36 certs
-rwx------ 1 root root  781 May 16 19:52 config.toml
</disable-copy>
{{</command>}}

You can see the file that contains all the runner configurations.

{{<command>}}
$ apt update && apt install nano
$ nano config.toml
{{</command>}}

The `privileged` field needs to be changed to `true`. Now the configurations should look like this:

```toml
connection_max_age = "15m0s"
shutdown_timeout = 0

[session_server]
  session_timeout = 1800

[[runners]]
  name = "localstack-testcontainers-runner"
  url = "https://gitlab.com"
  id = 36509569

>   token = "glrt-1SLQUHi3TjsEi23Ro48e"

Reviewer: Is it a real token? 👀

Author: it's a real token, but the runner does not exist anymore, so there's nothing you can do with it anymore. but I'll replace it with a generic pattern and explain it

  token_obtained_at = 2024-05-16T19:51:27Z

>   token_expires_at = 0001-01-01T00:00:00Z

Reviewer: IMHO Please remove such information

<br>
<br>
{{< figure src="coffee-app-diagram.png" width="80%" height="auto">}}

## CI Pipeline Using GitLab Runners

#### Configuring the test container

To follow along, make changes to the code or run your own pipelines, you can fork the repository at https://gitlab.com/tinyg210/coffee-backend-localstack.
<br>
The application is developed, built and tested locally, the next step is to establish a quality gate in the pipeline, to make sure nothing breaks. 

The basis for the container used for testing looks like this:

```java
@Testcontainers
public class LocalStackConfig {
    @Container
    protected static LocalStackContainer localStack =
            new LocalStackContainer(DockerImageName.parse("localstack/localstack-pro:3.4.0"))
                    .withEnv("LOCALSTACK_AUTH_TOKEN", System.getenv("LOCALSTACK_AUTH_TOKEN"))
                    .withCopyFileToContainer(
                            MountableFile.forHostPath("target/lambda.jar", 0777),
                            "/etc/localstack/init/ready.d/target/lambda.jar")
                    .withCopyFileToContainer(
                            MountableFile.forClasspathResource("init-resources.sh", 0777),

>                             "/etc/localstack/init/ready.d/init-resources.sh")

Author: tbh, i found the second one by exploring the library myself, looking for a suitable method. i also thought that adding more documentation links was out of scope. i think people can find the TC docs themselves if they really want to see extra configs.

all the necessary AWS resources. Both files are copied with read/write/execute permissions.
- `DEBUG=1` enables a more verbose debugging of LocalStack.
- `LAMBDA_DOCKER_FLAGS` sets specific Testcontainers labels to the Lambda containers, as a solution to be correctly managed by Ryuk.
Since the compute containers are created by LocalStack and not the Testcontainers framework, they do not receive the necessary tags.

{{< alert title="Sidenote" >}}

Ryuk is a component of Testcontainers that helps manage and clean up Docker resources created during testing. Specifically, Ryuk 
ensures that any Docker containers, networks, volumes, and other resources are properly removed when they are no longer needed. 
This prevents resource leaks and ensures that the testing environment remains clean and consistent between test runs.

When Testcontainers starts, it typically launches a Ryuk container in the background. This container continuously monitors
the Docker resources created by Testcontainers and removes them once the test execution is complete or if they are no longer in use.

> {{< /alert >}}

Author: fixed

useful for integration testing with Docker containers.
- Variables:
  - `DOCKER_HOST: tcp://docker:2375` - sets the Docker host to communicate with the Docker daemon inside the dind service.
  - `DOCKER_TLS_CERTDIR: ""` - we'll disable TLS to simplify the setup in a testing environment.
  - `DOCKER_DRIVER: overlay2` - specifies the storage driver for Docker, ensuring better performance and compatibility.
- The last line ensures that the pipeline fails if the tests fail.

#### Executors

We mentioned in the beginning that each job runs in a newly provisioned VM. You can also notice that the pipeline configuration mentions 
a docker image, which is a template that contains instructions for creating a container. This might look confusing, but a runner is responsible 
for the execution of one job. This runner is installed on a machine and implements 
a certain [executor](https://docs.gitlab.com/runner/executors/). The executor determines the environment in which the job runs. By
default, the GitLab-managed runners use a Docker Machine executor. Some other available executor options are: SSH, Shell, Parallels,
VirtualBox, Docker, Docker Autoscaler, Kubernetes.

> Sometimes visualising the components of a pipeline can be tricky, so let's simplify this into a diagram:

Reviewer 2: ```suggestion
Sometimes visualising the components of a pipeline can be tricky, so let's simplify this into a diagram:
```

Just noticed that you used British spelling, I think we are trying to write everything in US English. @HarshCasper please confirm/deny.  :)

Author: fixed

{{< figure src="gitlab-ci-diagram.png" width="80%" height="auto">}}

Basically, the `service` is an additional container that starts at the same time as the one running the `test_job`. The job container has
a Docker client, and it communicates with the Docker daemon, running in the service container, in order to spin up more containers, in this
case for the Lambda functions.

> Don't forget to add your LOCALSTACK_AUTH_TOKEN as a masked variable in your CI/CD settings. 

Reviewer 2: ```suggestion
Don't forget to add your `LOCALSTACK_AUTH_TOKEN` as a masked variable in your CI/CD settings. 
```

For simplicity, we'll use a GitLab Runner Docker image. The GitLab Runner Docker images are designed as wrappers around the standard
`gitlab-runner` command, like if GitLab Runner was installed directly on the host. You can read more about it in the [GitLab documentation](https://docs.gitlab.com/runner/install/docker.html).

Make sure you have Docker installed. To verify your setup you can run the `docker info` command.

Now, you need to create a volume on the disk that holds the configuration for the runner. You can have different volumes that can be
used for different runners.

{{<command>}}
$ docker volume create gitlab-runner-config
<disable-copy>
gitlab-runner-config
</disable-copy>
{{</command>}}

The agent that will interact with the system to create the runner needs to be run:

{{<command>}}
$ docker run -d --name gitlab-runner \
--restart always \
-v /var/run/docker.sock:/var/run/docker.sock \
-v gitlab-runner-config:/etc/gitlab-runner/ \
gitlab/gitlab-runner:latest
<disable-copy>
79ad150847bd78fc567b08cafd96fc6bfec8f1946feb2beea89d7c6c395c01c4
</disable-copy>
{{</command>}}

The breakdown for this command:

> - The container is named gitlab-runner.

Reviewer 2: ```suggestion
- The container is named `gitlab-runner`.
```

-rw------- 1 root root   14 May 16 19:50 .runner_system_id
drwx------ 2 root root 4096 May  3 17:36 certs
-rwx------ 1 root root  781 May 16 19:52 config.toml
</disable-copy>
{{</command>}}

You can see the file that contains all the runner configurations.

{{<command>}}
$ apt update && apt install nano
$ nano config.toml
{{</command>}}

The `privileged` field needs to be changed to `true`. Now the configurations should look like this:

```toml
connection_max_age = "15m0s"
shutdown_timeout = 0

[session_server]
  session_timeout = 1800

[[runners]]
  name = "localstack-testcontainers-runner"
  url = "https://gitlab.com"
  id = 36509569
  token = "glrt-1SLQUHi3TjsEi23Ro48e"
  token_obtained_at = 2024-05-16T19:51:27Z
  token_expires_at = 0001-01-01T00:00:00Z
  executor = "docker"
  [runners.custom_build_dir]
  [runners.cache]
    MaxUploadedArchiveSize = 0
    [runners.cache.s3]
    [runners.cache.gcs]
    [runners.cache.azure]
  [runners.docker]
    tls_verify = false
    image = "docker:26.1.2-dind"
    privileged = true
    disable_entrypoint_overwrite = false
    oom_kill_disable = false
    disable_cache = false
    volumes = ["/cache"]
    shm_size = 0
    network_mtu = 0
```

> `Control + X` to save and exit the file. The runner is ready to use. You can now run your pipeline by pushing changes to your project 

Reviewer 2: ```suggestion
`[CTRL] + [X]` to save and exit the file. The runner is ready to use. You can now run your pipeline by pushing changes to your project 
```

Just a suggestion. For me, this is easier to understand.
